lights/light_regulator/LampRegulator.cs

```diff
@@ -0,0 +1,138 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+
+namespace Lamp_Regulator 
+{
+    class Regulator
+    {
+        private byte[] _buffer = new byte[1024];
+        private List<Socket> _clientSockets = new List<Socket>();
+        private Socket _serverSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+        private int BACKLOG;
+        private int PORT;
+
+
+        public Regulator(int port, int backlog)
+        {
+            BACKLOG = backlog;
+            PORT = port;
+        }
+
+        public void StartRegulator()
+        {
+            SetupServer();
+        }
+
+        public void CloseRegulator()
+        {
+            CloseAllSockets();
+        }
+
+        private void CloseAllSockets()
+        {
+            foreach (Socket socket in _clientSockets)
+            {
+                socket.Shutdown(SocketShutdown.Both);
+                socket.Close();
+            }
+
+            _serverSocket.Close();
+        }
+
+        private void SetupServer()
+        {
+            Console.WriteLine("Setting up server ...");
+            _serverSocket.Bind(new IPEndPoint(IPAddress.Any, PORT));
+            _serverSocket.Listen(BACKLOG);
+            _serverSocket.BeginAccept(new AsyncCallback(AcceptCallback), null);
+            Console.Read();
+
+        }
+
+        private void AcceptCallback(IAsyncResult AR)
+        {
+            Socket socket;
+            try
+            {
+                socket = _serverSocket.EndAccept(AR);
+            }
+            catch (ObjectDisposedException)
+            {
+                return;
+            }
+
+            byte[] buffer = new byte[1024];
+            Console.WriteLine("client connected");
+            _clientSockets.Add(socket);
+            socket.BeginReceive(buffer, 0, buffer.Length, SocketFlags.None, ReceiveCallback, new SocketMemory(socket, buffer));
+            _serverSocket.BeginAccept(new AsyncCallback(AcceptCallback), null);
+        }
+
+        private void ReceiveCallback(IAsyncResult AR)
+        {
+            SocketMemory memory = (SocketMemory)AR.AsyncState;
+            Socket socket = memory.socket; 
+
+            int received;
+
+            try
+            {
+                received = socket.EndReceive(AR);
+            }
+            catch (SocketException)
+            {
+                Console.WriteLine("client forcefully disconnected");
+                socket.Close();
+                _clientSockets.Remove(socket);
+                return;
+            }
+
+            byte[] dataBuf = new byte[received];
```

Comment:
nie jestem pewien po co tworzymy tutaj `dataBuf`. Po to, żeby później od razu na tym użyć `Encoding.ASCII.GetString(dataBuf)`? Równie dobrze można by napisać 
```c#
string text = Encoding.ASCII.GetString(memory.buffer, 0, received);
```